feat(anoncreds): add anoncreds API

TimoGlastra opened the pull request.

TimoGlastra commented:
revocationList -> revocationStatusList

genaris commented on packages/anoncreds/src/services/AnonCredsHolderService.ts:
```diff
@@ -19,8 +21,10 @@ export interface AnonCredsHolderService {
+  // now I think it's not that much of an issue
+  getCredential(agentContext: AgentContext, options: GetCredentialOptions): Promise<AnonCredsCredentialInfo>
 
   createCredentialRequest(
```
What about MasterSecret creation and storage? It is part of AnonCredsApi, isn't it? If that's the case, will it be passed in `CreateCredentialRequestOptions` and `CreateProofOptions`?

TimoGlastra replied:
Yeah it should be handled by the AnonCreds Api / services (prob the holder service), and removed from the wallet interface.

genaris replied:
I figured out that it's only needed for CreateCredentialRequestOptions and can even be optional, using a default one as in 99% of the cases it will be a single master secret. I've implemented some draft classes in my PR https://github.com/TimoGlastra/aries-framework-javascript/pull/18

genaris reviewed (commented):
I know this is still a draft but I've been checking it for an integration with anoncreds-rs and I left some comments, mostly some questions to make sure I understood correctly the way these interfaces should be used.

TimoGlastra commented on packages/anoncreds/src/services/AnonCredsIssuerService.ts:
```diff
@@ -9,6 +9,8 @@ import type { AnonCredsCredentialOffer } from '../models/exchange'
+
 export interface AnonCredsIssuerService {
   createSchema(agentContext: AgentContext, options: CreateSchemaOptions): Promise<AnonCredsSchema>
 
```
My idea is that we add records for this in the AnonCreds package and we instruct that the same repositories are used by implementations that can use it. However, for Indy SDK we can't use the records and therefore it's part of the specific issuer/holder/verifier services.

This way there also won't be migration scripts needed, as they'll be using the same records. The idea is to create a separate record for the private parts so it's not exposed to the end user (unless they specifically fetch the private record).

See https://github.com/hyperledger/aries-framework-javascript/issues/1193 and https://github.com/hyperledger/aries-framework-javascript/issues/1164 for extra context.

Do you think this approach will work? Happy to go other directions

genaris replied:
Yes, I think it will be better that way, especially because we are talking about records that are aligned with AnonCreds specification, so any implementation should be fine by using it (and if it needs something else, it can probably use an internal metadata value).

The same would apply to `AnonCredsCredentialRecord`, right? I mean there will be a common Record/Repository and each `AnonCredsHolderService` implementation will use it. In such case, wouldn't be the methods `getCredential` / `deleteCredential` / `storeCredential` nearly the same for every implementation (except of indy-sdk of course)?

TimoGlastra replied:
Yes probably! I think `storeCredential` in indy-sdk verifies the credential so that will be different.

TimoGlastra replied:
```
interface AnonCredsCredentialInfo {
        credentialId: string
        attributes: Record<string, string>
        schemaId: string
        credentialDefinitionId: string
        revocationRegistryId?: string
        credentialRevocationId?: string
}

getCredential(credentialId: string): Promise<AnonCredsCredentialInfo>
deleteCredential(credentialId: string): Promise<void>
```

TimoGlastra replied:
```
interface AnonCredsCredentialInfo {
        credentialId: string
        attributes: Record<string, string>
        schemaId: string
        credentialDefinitionId: string
        revocationRegistryId?: string
        credentialRevocationId?: string
}

getCredential(credentialId: string): Promise<AnonCredsCredentialInfo>
deleteCredential(credentialId: string): Promise<void>
```

TimoGlastra commented:
> I know this is still a draft but I've been checking it for an integration with anoncreds-rs and I left some comments, mostly some questions to make sure I understood correctly the way these interfaces should be used.

Thanks for the feedback, comments!

TimoGlastra commented:
> Not completely related with this PR but it seems that `packages/anoncreds/src/formats/__tests__/LegacyIndyCredentialFormatService.test.ts` takes quite a significant time in CI (between 20 and 30 seconds). Not sure where it is getting stuck but maybe we'll need to increase the timeout for anoncreds suite.

It's actually doing integration tests (while the indy-sdk was mocked before) so that's prob what's taking so long. I've been a bit more focused on the integration tests as unit tests take A LOT of time to write and often don't really test if something actually works. But it does mean tests take longer to run ....

TimoGlastra commented:
This is now ready for review. I've renamed master secret to link secret and opened a new issue for adding the getCredential/getCredentials methods.

I've added end to end tests for the functionality in anoncreds.test.ts which uses an in memory anoncreds registry to register/resolve all anoncreds objects

genaris commented on packages/anoncreds/src/services/AnonCredsHolderServiceOptions.ts:
```diff
@@ -81,9 +81,19 @@ export type GetCredentialsForProofRequestReturn = Array<{
 export interface CreateCredentialRequestOptions {
   credentialOffer: AnonCredsCredentialOffer
   credentialDefinition: AnonCredsCredentialDefinition
+  masterSecretId?: string
```
```suggestion
  linkSecretId?: string
```

TimoGlastra replied:
I've applied the suggestion. It's currently not used in the IndySdkHolderService (it's using the masterSecretId from the wallet).

However, I think we can best address this in #1198

genaris commented on packages/anoncreds/tests/InMemoryAnonCredsRegistry.ts:
```diff
@@ -26,8 +28,28 @@ export class InMemoryAnonCredsRegistry implements AnonCredsRegistry {
+  private schemas: Record<string, AnonCredsSchema>
+  private credentialDefinitions: Record<string, AnonCredsCredentialDefinition>
+  private revocationRegistryDefinitions: Record<string, AnonCredsRevocationRegistryDefinition>
+  // { revocationRegistryDefinitionId: { timestamp: revocationStatusList }
```
Is this commented code left for some purpose to handle later on?

TimoGlastra replied:
It can be removed! 👍

genaris reviewed (changes requested):
I left some very minor comments. There are a few more details (like defining KeyCorrectnessProof and CredentialDefinitionPrivate structure) but we can handle in another PR.

TimoGlastra commented on packages/anoncreds/tests/InMemoryAnonCredsRegistry.ts:
```diff
@@ -26,8 +28,28 @@ export class InMemoryAnonCredsRegistry implements AnonCredsRegistry {
+  private schemas: Record<string, AnonCredsSchema>
+  private credentialDefinitions: Record<string, AnonCredsCredentialDefinition>
+  private revocationRegistryDefinitions: Record<string, AnonCredsRevocationRegistryDefinition>
+  // { revocationRegistryDefinitionId: { timestamp: revocationStatusList }
```
```suggestion
```

genaris commented:
Seems that passing anoncreds test is still hard due to timing issues... Based on the run for node 14 it took around 23 secs to pass. Maybe would be neeeded to increase the timeout to 30000?
